docs/sphinx/dev/dev_style_guide_code.rst

```diff
@@ -1,28 +1,169 @@
-Code Guidelines
------------------
+C++ Code Guidelines
+---------------------
 
-Classes
+Naming
 ^^^^^^^^^^
-Classes in kotekan should use *CamelCase* formatting, e.g. ``myFavouriteProcess``.
+- Type names in kotekan should be nouns, use *CamelCase* formatting and begin
+  with a capital letter, e.g. ``MyFavouriteProcess``. This includes classes,
+  enums, structs and typedefs.
+- With the exception of class constructors & destructors, function names should
+  use underscore notation and begin with a lower case letter, e.g. ``my_func``.
 
-Functions
+Variables
 ^^^^^^^^^^
-With the exception of class constructors & destructors, function names should use underscore notation,
-e.g. ``my_func``.
+Variables in the code should use underscore naming, e.g.
+``my_favourite_variable``.
 
+Explicit typing should be used wherever possible, e.g. always use ``uint32_t``
+rather than ``uint``.
 
-Variables
+Private member variables should start with an underscore, e.g.
+``_my_private_variable``.
+
+Namespaces
 ^^^^^^^^^^
-Variables in the code should use underscore naming, e.g. ``my_favourite_variable``.
+Avoid `using namespace X;`. Instead specify where you are using classes or
+functions from a namespace, e.g. `std::vector<std::string> my variable;`.
+Never do `using namespace std;`.
 
-Explicit typing should be used wherever possible, e.g. always use ``uint32_t`` rather than ``uint``.
+Header files
+^^^^^^^^^^^^^
+C++ header files should be named after the class they describe and use the
+ending `.hpp`. They should begin with a header file commit like
 
+.. code-block:: c++
 
-Variables
-^^^^^^^^^^
+    /****************************************************
+    @file
+    @brief Brief description of the classes in this file.
+    - MyClass
+    - MyOtherClass : public MyOtherBaseClass
+    *****************************************************/
 
-Structs
-^^^^^^^^^^
+Immediately after the header file comment include guards should follow:
 
-Enums
-^^^^^^^^^^
+.. code-block:: c++
+
+    #ifndef MY_CLASS_H
+    #define MY_CLASS_H
+
+With a simple
+
+.. code-block:: c++
+
+    #endif
+
+at the end of the file.
+
+
+#include Style
+^^^^^^^^^^^^^^^
+The list of #includes should be in the order:
+
+ - Header file implementing the interface of this `.cpp` file.
```

Comment:
This order doesn't quite make sense to me. I'd suggest:
- Interface file
- Local/private
- Project
- System

as it's most to least specific. The suggestion doesn't seem to have a consistent order. Personally I'd usually do the other way (i.e. least to most), but that's Python influence I think.

Reply:
Richard's order makes sense to me too, not least because it helps me make a mental model of the module dependency hierarchy. 

This is also the order recommended in "Thinking in C++" (via [SO](https://stackoverflow.com/a/14243688)), although the [Google style guide](http://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) suggests the reverse:

> 1. dir2/foo2.h.
> 2. A blank line
> 3. C system files.
> 4. C++ system files.
> 5. A blank line
> 6. Other libraries' .h files.
> 7. Your project's .h files.

Technically, either is OK, as long as the interface file is first, because that helps catch errors with missing includes in the interface header.

we should emphasize that within each section, includes should be sorted alphabetically.